Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm

```diff
@@ -244,6 +244,79 @@
+{
+    RefPtr destinationExtension = extensionController()->extensionContext(extensionID);
+    if (!destinationExtension) {
+        completionHandler(std::nullopt, @"extension not found, but do not report this error");
+        return;
+    }
+
+    WebExtensionMessageSenderParameters completeSenderParameters = senderParameters;
+    if (RefPtr tab = getTab(senderParameters.pageProxyIdentifier))
+        completeSenderParameters.tabParameters = tab->parameters();
+
+    auto mainWorldProcesses = processes(WebExtensionEventListenerType::RuntimeOnMessageExternal, WebExtensionContentWorldType::Main);
+    if (mainWorldProcesses.isEmpty()) {
+        completionHandler(std::nullopt, std::nullopt);
+        return;
+    }
+
+    auto callbackAggregator = EagerCallbackAggregator<ReplyCompletionHandlerSignature>::create(WTFMove(completionHandler), std::nullopt, std::nullopt);
+
+    for (auto& process : mainWorldProcesses) {
+        process->sendWithAsyncReply(Messages::WebExtensionContextProxy::DispatchRuntimeMessageEvent(WebExtensionContentWorldType::Main, messageJSON, std::nullopt, completeSenderParameters), [callbackAggregator](std::optional<String> replyJSON) {
+            callbackAggregator.get()(replyJSON, std::nullopt);
+        }, identifier());
+    }
+}
+
+void WebExtensionContext::runtimeWebPageConnect(const String& extensionID, WebExtensionPortChannelIdentifier channelIdentifier, const String& name, const WebExtensionMessageSenderParameters& senderParameters, CompletionHandler<void(WebExtensionTab::Error)>&& completionHandler)
+{
+    RefPtr destinationExtension = extensionController()->extensionContext(extensionID);
+    if (!destinationExtension) {
+        completionHandler(@"extension not found, but do not report this error");
+        return;
+    }
+
+    WebExtensionMessageSenderParameters completeSenderParameters = senderParameters;
+    if (RefPtr tab = getTab(senderParameters.pageProxyIdentifier))
+        completeSenderParameters.tabParameters = tab->parameters();
+
+    auto validMatchPatterns = destinationExtension->extension().externallyConnectableMatchPatterns();
+    if (!WebExtensionMatchPattern::matchPatternSetContainsMatchForURL(validMatchPatterns, completeSenderParameters.url)) {
+        completionHandler(@"extension does not have access to this page, but do not reprort this error");
+        return;
+    }
+
+    constexpr auto sourceContentWorldType = WebExtensionContentWorldType::WebPage;
+    constexpr auto targetContentWorldType = WebExtensionContentWorldType::Main;
+
+    // Add 1 for the starting port here so disconnect will balance with a decrement.
+    addPorts(sourceContentWorldType, channelIdentifier, 1);
+
+    auto mainWorldProcesses = processes(WebExtensionEventListenerType::RuntimeOnConnectExternal, targetContentWorldType);
+    if (mainWorldProcesses.isEmpty()) {
+        completionHandler(toErrorString(@"runtime.connect()", nil, @"no runtime.onConnectExternal listeners found"));
```

Comment:
@xeenon is this an error that we actually do want to report?

Reply:
I don't think so. That should be just like the extension does not exist to the page.